cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowUtil.qll

```diff
@@ -520,6 +521,22 @@ predicate localFlowStep(Node nodeFrom, Node nodeTo) {
   FieldFlow::fieldFlow(nodeFrom, nodeTo)
 }
 
+private predicate pointerWrapperFlow(Node nodeFrom, Node nodeTo) {
+  // post-update-smart-pointer-`operator->` -> `post-update`-qualifier
+  exists(PointerWrapper wrapper, Call call |
+    call = wrapper.getAnUnwrapperFunction().getACallToThisFunction() and
+    nodeFrom.(PostUpdateNode).getPreUpdateNode().asExpr() = call and
+    nodeTo.asDefiningArgument() = call.getQualifier()
+  )
+  or
+  // smart-pointer-qualifier -> smart-pointer-`operator->`
+  exists(PointerWrapper wrapper, Call call |
+    call = wrapper.getAnUnwrapperFunction().getACallToThisFunction() and
+    nodeFrom.asExpr() = call.getQualifier() and
+    nodeTo.asExpr() = call
+  )
```

Comment:
Am I right that this entry adds data flow from `mySmartPointer` to `*mySmartPointer`? I don't see how that's justified, given that we don't have data flow from `myRawPointer` to `*myRawPointer` either. We do want taint flow, though.

Reply:
Hm, yes. That's a good point. We don't want that flow here. We do want flow in the case of `operator->` and `get` on smart pointers, but not on `operator*`. I guess this goes back to the discussion from [here](https://github.com/github/codeql/pull/5543#discussion_r602358895) of whether we should have two predicates on `PointerWrapper`. We could also just specifically exclude `operator*` here.

Reply:
Rather than syntactically distinguishing `->` from `*`, perhaps allow the flow only if the called function returns a pointer rather than a reference.

Reply:
Fixed in 80d5b17900ebf200de067e0a7cc7466606bfea7b. Note that this removes many of the results that use pointer dereferences. We get back these results when we merge https://github.com/github/codeql/pull/5578 into this PR (due to https://github.com/github/codeql/pull/5578/files#diff-b5712695aa56fbc8e53df62819c2c9173d658a28cd4ea57ff322ba6db9adbc15R145).